src/waifu2x.rs

```diff
@@ -40,65 +53,119 @@ impl Waifu2x {
     pub fn init() {
         unsafe { init_ncnn() }
     }
-    pub fn new(gpuid: u8, noise: u8, scale: u8, tilesize: u16, is_cunet: bool) -> Self {
+
+    pub fn new(
+        gpuid: u8,
+        noise: i8,
+        scale: u8,
+        tilesize: u16,
+        is_cunet: bool,
+    ) -> Result<Self, Waifu2xError> {
+        // valid noise values: -1,0,1,2,3
+        if !(-1..=3).contains(&noise) {
+            return Err(Waifu2xError::InvalidNoise(noise));
+        }
+
+        // check if scale is 0, is not power of 2, or more than 32
+        // valid values are: 1,2,4,8,16,32
+        if scale == 0 || scale & (scale - 1) != 0 || scale > 32 {
+            return Err(Waifu2xError::InvalidScale(scale));
+        }
+
+        // check whether gpuid is valid or not
+        let gpu_count = Self::get_gpu_count();
+        // valid values: 0..gpu_count
+        if !(0..gpu_count).contains(&gpuid.into()) {
+            return Err(Waifu2xError::InvalidGpuId(gpuid));
+        }
+
         unsafe {
             let config = init_config(
                 i32::from(noise),
-                i32::from(scale),
+                // this always scales by 2 at a time
+                2,
```

Comment:
waifu2x has a model that only reduces noise but does not scale: https://github.com/darkskygit/waifu2x.rs/tree/models/models-cunet

Reply:
Are there any features or ways I can test for this specific case? In other instances, this scale must be 2, or the resulting image will be corrupted

Reply:
for example, you can use this model to denoise an image while keeping the size unchanged: https://github.com/darkskygit/waifu2x.rs/blob/models/models-cunet/noise0_model.bin
in this case, we can use the model by passing in scale=1
this model can be used to repair [artifacts](https://en.wikipedia.org/wiki/Compression_artifact) generated by jpeg/webp

Reply:
waifu2x is not actually a model that takes an input of m*n array and outputs an m2*n2 array
it is a model that enhances the details of an already enlarged image
therefore, the enlargement process is a separate step. In this implementation, you can see the specific implementation here: https://github.com/darkskygit/waifu2x.rs/blob/913be4b7770cab00423cb79a43e9b3041b96c5df/waifu2x/waifu2x.cpp#L381-L392

Reply:
Is adding a case for `#[cfg(all(feature = "model_bundled", feature = "noise"))]` and forcing `scale = 1` sufficient to solve this?

Reply:
in the current implementation, we can choose different models by passing in `scale 1-2` and `noise 0-3` respectively
each combination has a corresponding model in waifu2x: https://github.com/darkskygit/waifu2x.rs/tree/models/models-cunet
do you think it would be difficult to maintain the same functionality in the current pull request?

Reply:
> do you think it would be difficult to maintain the same functionality in the current pull request?

Not at all. If I understand you correctly, the internal scale should be `1` if user passes in `1`, else a user scale of `>= 2` will set it to internal scaling of `2` (which if `> 2` will also do additional runs to multiply the scale).

I adjusted the code to do the following. Is this the logic you were getting at?
| Noise Level | User Scale | Internal Scale Used | Additional Runs* |
|-------------|------------|---------------------|----------------|
| 0           |      1x      |          1x           | 0 |
| 0           |      2x      |          2x           | 0 |
| 0           |      4x      |          2x           | 1 |
| 0           |      8x      |          2x           | 2 |
| 0           |      16x      |        2x             | 3 |
| 0           |      32x      |        2x             | 4 |
| 1           |      1x      |          1x           | 0 |
| 1           |      2x      |          2x           | 0 |
| 1           |      4x      |          2x           | 1 |
| 1           |      8x      |          2x           | 2 |
| 1           |      16x      |        2x             | 3 |
| 1           |      32x      |        2x             | 4 |
| 2           |      1x      |          1x           | 0 |
| 2           |      2x      |          2x           | 0 |
| 2           |      4x      |          2x           | 1 |
| 2           |      8x      |          2x           | 2 |
| 2           |      16x      |        2x             | 3 |
| 2           |      32x      |        2x             | 4 |
| 3           |      1x      |          1x           | 0 |
| 3           |      2x      |          2x           | 0 |
| 3           |      4x      |          2x           | 1 |
| 3           |      8x      |          2x           | 2 |
| 3           |      16x      |        2x             | 3 |
| 3           |      32x      |        2x             | 4 |

\* 1 initial 2x run is always done (downsizing if needed if scale was 1x, otherwise leaves it at 2x), then additional runs 2x are done as many times as needed to multiply up to the user scale (4x user scale = 1 additional run, for 2x * 2x = 4x total)

Reply:
One thing I'm seeing from the above code, is that an internal scale `> 2` will hit the default of `return &scale2_0x_model_bin[0];` which will ignore whatever noise a user specifies. (So say I wanted to upscale to 4x, with a noise of 3, it is impossible).

With the way I currently have the code set up, it is either 1 or 2, so it always respects the noise level, however that default switch case will never be reached.

Your thoughts on this?

Reply:
> I'd like to re-introduce noise level `-1` as a way to disable it, but at the moment I have no real internal api to achieve the disabling of the noise model. (And honestly, I don't know C++, minus the very basic edit I did before)

i think it's appropriate
`scale2_0x_model_bin` in the official design, it means that the details are enhanced without the repair of artifacts, which is indeed equivalent to not turning on noise reduction for lossless pictures

Reply:
Let me know if this looks good for you now, or if you'd like any other changes 😁

Reply:
LGTM!